File: apps/ios/Sources/Core/Seed.swift
import Foundation

// JavaScript's bitwise operators work on 32-bit signed integers.
// Swift's Int is 64-bit, so we must explicitly truncate to Int32
// to match JS behavior exactly. This is the most critical function
// for cross-platform parity — if hashes diverge, every puzzle differs.
func simpleHash(_ str: String) -> Int {
    var hash: Int32 = 0
    for scalar in str.unicodeScalars {
        let char = Int32(scalar.value)
        // JS: hash = ((hash << 5) - hash) + char; hash = hash & hash;
        // The (hash << 5) - hash is hash * 31, but we must use bitwise
        // shift to match JS overflow behavior exactly.
        hash = (hash &<< 5) &- hash &+ char
        // JS "hash & hash" is a no-op identity that forces i32 truncation.
        // In Swift with Int32, this is already the case — no extra op needed.
    }
    return Int(abs(hash))
}

// ── Deck dealing (§215) — port of packages/core seed.ts ─────────────────────
// Post-cutover daily seeds deal from a deterministic shuffled deck (one
// Fisher–Yates permutation per epoch, fixed non-overlapping per-mode slices
// per day) so no answer repeats across ANY 5-letter daily for a whole epoch
// (~40 days). Pre-cutover dailies, unlimited/match/replacement seeds keep the
// legacy hash path. Byte-parity with TS/Kotlin is pinned by seed-fixtures.json.
public let DECK_CUTOVER_DATE = "2026-08-24"

private struct DeckDeal { let offset: Int; let count: Int }

// FROZEN deal table — never reorder/resize existing modes (see TS core).
private let DECK_DEALS_5: [String: DeckDeal] = [
    "DUEL": DeckDeal(offset: 0, count: 1),
    "QUORDLE": DeckDeal(offset: 1, count: 4),
    "OCTORDLE": DeckDeal(offset: 5, count: 8),
    "SEQUENCE": DeckDeal(offset: 13, count: 4),
    "RESCUE": DeckDeal(offset: 17, count: 4),
    "GAUNTLET": DeckDeal(offset: 21, count: 21),
    "DUEL_VS": DeckDeal(offset: 42, count: 1),
]
private let DECK_TOTAL_5 = 43
private let DECK_DEALS_6: [String: DeckDeal] = ["DUEL_6": DeckDeal(offset: 0, count: 1)]
private let DECK_DEALS_7: [String: DeckDeal] = ["DUEL_7": DeckDeal(offset: 0, count: 1)]

/// mulberry32 PRNG — UInt32 wrapping ops reproduce the JS int32/imul bit
/// patterns exactly. As parity-critical as simpleHash.
private struct Mulberry32 {
    var state: UInt32
    mutating func next() -> UInt32 {
        state = state &+ 0x6D2B79F5
        var t = (state ^ (state >> 15)) &* (state | 1)
        t = (t &+ ((t ^ (t >> 7)) &* (t | 61))) ^ t
        return t ^ (t >> 14)
    }
}

/// Whole days from DECK_CUTOVER_DATE to `date` (UTC), nil if unparseable.
private func deckDayIndex(_ date: String) -> Int? {
    let f = DateFormatter()
    f.dateFormat = "yyyy-MM-dd"
    f.timeZone = TimeZone(identifier: "UTC")
    f.locale = Locale(identifier: "en_US_POSIX")
    guard let to = f.date(from: date), let from = f.date(from: DECK_CUTOVER_DATE) else { return nil }
    return Int(((to.timeIntervalSince1970 - from.timeIntervalSince1970) / 86400).rounded())
}

private func deckSolutions(_ seed: String, count: Int, pool: [String],
                           deals: [String: DeckDeal], dailyTotal: Int, lengthKey: Int) -> [String]? {
    // Exactly daily-YYYY-MM-DD-MODE: suffixed seeds (gauntlet blackout
    // replacements, etc.) must never collide with the day's real deals.
    let parts = seed.split(separator: "-", omittingEmptySubsequences: false)
    guard parts.count == 5, parts[0] == "daily" else { return nil }
    let date = "\(parts[1])-\(parts[2])-\(parts[3])"
    guard date >= DECK_CUTOVER_DATE else { return nil }
    guard let deal = deals[String(parts[4])], deal.count == count else { return nil }
    let daysPerEpoch = pool.count / dailyTotal
    guard daysPerEpoch >= 1 else { return nil }
    guard let dayIndex = deckDayIndex(date), dayIndex >= 0 else { return nil }
    let epoch = dayIndex / daysPerEpoch
    var perm = Array(0..<pool.count)
    var rng = Mulberry32(state: UInt32(truncatingIfNeeded: simpleHash("deck-\(lengthKey)-\(epoch)")))
    var i = pool.count - 1
    while i > 0 {
        let j = Int(rng.next() % UInt32(i + 1))
        perm.swapAt(i, j)
        i -= 1
    }
    let base = (dayIndex % daysPerEpoch) * dailyTotal + deal.offset
    return (0..<count).map { pool[perm[base + $0]] }
}
// ────────────────────────────────────────────────────────────────────────────

public func generateSolutionsFromSeed(_ seed: String, count: Int) -> [String] {
    let dict = GameDictionary.shared
    // Answer pool chosen by the DATE embedded in the seed (never wall clock),
    // so every client resolves the same seed identically: pre-cutover daily
    // dates → legacy list, post-cutover dailies + all non-daily seeds → curated.
    let pool = dict.solutionPool(forDateKey: getDailySeedDate(seed))
    // Post-DECK_CUTOVER_DATE dailies deal from the shuffled deck (§215).
    if let deck = deckSolutions(seed, count: count, pool: pool,
                                deals: DECK_DEALS_5, dailyTotal: DECK_TOTAL_5, lengthKey: 5) {
        return deck
    }
    let solutionCount = pool.count
    var solutions: [String] = []
    var used: Set<Int> = []

    for i in 0..<count {
        let seedWithIndex = "\(seed)-\(i)"
        var hash = simpleHash(seedWithIndex)
        var attempts = 0

        while used.contains(hash % solutionCount) && attempts < solutionCount {
            hash = simpleHash("\(seedWithIndex)-\(attempts)")
            attempts += 1
        }

        let index = hash % solutionCount
        used.insert(index)
        solutions.append(pool[index])
    }

    return solutions
}

public func generateSolutionsFromSeedForLength(_ seed: String, count: Int, wordLength: Int) -> [String] {
    let dict = GameDictionary.shared
    // Same date-gate as the 5-letter path — pre-cutover Six/Seven dailies keep
    // their legacy words; new dailies + non-daily seeds use the curated list.
    let pool = dict.solutionPool(forLength: wordLength, dateKey: getDailySeedDate(seed))
    // Six/Seven post-cutover dailies deal from their own decks (§215).
    if let deals = wordLength == 6 ? DECK_DEALS_6 : wordLength == 7 ? DECK_DEALS_7 : nil,
       let deck = deckSolutions(seed, count: count, pool: pool,
                                deals: deals, dailyTotal: 1, lengthKey: wordLength) {
        return deck
    }
    let solutionCount = pool.count
    var solutions: [String] = []
    var used: Set<Int> = []

    for i in 0..<count {
        let seedWithIndex = "\(seed)-\(i)"
        var hash = simpleHash(seedWithIndex)
        var attempts = 0

        while used.contains(hash % solutionCount) && attempts < solutionCount {
            hash = simpleHash("\(seedWithIndex)-\(attempts)")
            attempts += 1
        }

        let index = hash % solutionCount
        used.insert(index)
        solutions.append(pool[index])
    }

    return solutions
}

public func generateMatchSeed() -> String {
    return "\(Int(Date().timeIntervalSince1970 * 1000))-\(UUID().uuidString.prefix(15).lowercased())"
}

public func generateDailySeed(date: String, gameMode: String) -> String {
    return "daily-\(date)-\(gameMode)"
}

public func isDailySeed(_ seed: String) -> Bool {
    return seed.hasPrefix("daily-")
}

public func getDailySeedDate(_ seed: String) -> String? {
    guard isDailySeed(seed) else { return nil }
    let parts = seed.split(separator: "-")
    guard parts.count >= 4 else { return nil }
    return "\(parts[1])-\(parts[2])-\(parts[3])"
}

File: apps/android/core/src/main/kotlin/com/wordocious/core/Seed.kt
package com.wordocious.core

import kotlin.math.abs

/**
 * JS-compatible 32-bit rolling hash. Kotlin `Int` is 32-bit and wraps on
 * overflow (two's complement), which exactly matches JavaScript's `hash & hash`
 * i32 truncation and the Swift port's `&<<` / `&-` / `&+`.
 *
 * THE most parity-critical function in the engine: if this diverges, every
 * daily puzzle differs from web/iOS. Validated against the shared
 * `hash-fixtures.json` (identical vectors to the TS and Swift suites).
 *
 * Mirrors `apps/ios/Sources/Core/Seed.swift` `simpleHash` and
 * `packages/core` `simpleHash`.
 */
fun simpleHash(str: String): Int {
    var hash = 0
    for (ch in str) {
        // JS: hash = ((hash << 5) - hash) + charCode; hash = hash & hash;
        // Kotlin Int already wraps to 32 bits, so no explicit mask is needed.
        hash = (hash shl 5) - hash + ch.code
    }
    return abs(hash)
}

/** `"daily-{date}-{gameMode}"` — date is local `yyyy-MM-dd`. */
fun generateDailySeed(date: String, gameMode: String): String = "daily-$date-$gameMode"

fun isDailySeed(seed: String): Boolean = seed.startsWith("daily-")

fun getDailySeedDate(seed: String): String? {
    if (!isDailySeed(seed)) return null
    val parts = seed.split("-")
    if (parts.size < 4) return null
    return "${parts[1]}-${parts[2]}-${parts[3]}"
}

// ── Deck dealing (§215) — port of packages/core seed.ts ─────────────────────
// Post-cutover daily seeds deal from a deterministic shuffled deck (one
// Fisher–Yates permutation per epoch, fixed non-overlapping per-mode slices
// per day) so no answer repeats across ANY 5-letter daily for a whole epoch
// (~40 days). Pre-cutover dailies, unlimited/match/replacement seeds keep the
// legacy hash path. Byte-parity with TS/Swift is pinned by seed-fixtures.json.
const val DECK_CUTOVER_DATE = "2026-08-24"

private data class DeckDeal(val offset: Int, val count: Int)

// FROZEN deal table — never reorder/resize existing modes (see TS core).
private val DECK_DEALS_5 = mapOf(
    "DUEL" to DeckDeal(0, 1),
    "QUORDLE" to DeckDeal(1, 4),
    "OCTORDLE" to DeckDeal(5, 8),
    "SEQUENCE" to DeckDeal(13, 4),
    "RESCUE" to DeckDeal(17, 4),
    "GAUNTLET" to DeckDeal(21, 21),
    "DUEL_VS" to DeckDeal(42, 1),
)
private const val DECK_TOTAL_5 = 43
private val DECK_DEALS_6 = mapOf("DUEL_6" to DeckDeal(0, 1))
private val DECK_DEALS_7 = mapOf("DUEL_7" to DeckDeal(0, 1))

/** mulberry32 PRNG — Kotlin Int wraps to 32 bits, so `+`/`*` reproduce the JS
 *  int32/imul bit patterns exactly. As parity-critical as [simpleHash]. */
private class Mulberry32(private var state: Int) {
    /** Next value as an unsigned 32-bit quantity in a Long (for `% n`). */
    fun nextU32(): Long {
        state += 0x6D2B79F5
        var t = (state xor (state ushr 15)) * (state or 1)
        t = (t + ((t xor (t ushr 7)) * (t or 61))) xor t
        return (t xor (t ushr 14)).toLong() and 0xFFFFFFFFL
    }
}

/** Whole days from DECK_CUTOVER_DATE to `date` (UTC), null if unparseable. */
private fun deckDayIndex(date: String): Int? {
    return try {
        val from = java.time.LocalDate.parse(DECK_CUTOVER_DATE)
        val to = java.time.LocalDate.parse(date)
        java.time.temporal.ChronoUnit.DAYS.between(from, to).toInt()
    } catch (_: Exception) {
        null
    }
}

private fun deckSolutions(
    seed: String, count: Int, pool: List<String>,
    deals: Map<String, DeckDeal>, dailyTotal: Int, lengthKey: Int,
): List<String>? {
    // Exactly daily-YYYY-MM-DD-MODE: suffixed seeds (gauntlet blackout
    // replacements, etc.) must never collide with the day's real deals.
    val parts = seed.split("-")
    if (parts.size != 5 || parts[0] != "daily") return null
    val date = "${parts[1]}-${parts[2]}-${parts[3]}"
    if (date < DECK_CUTOVER_DATE) return null
    val deal = deals[parts[4]] ?: return null
    if (deal.count != count) return null
    val daysPerEpoch = pool.size / dailyTotal
    if (daysPerEpoch < 1) return null
    val dayIndex = deckDayIndex(date) ?: return null
    if (dayIndex < 0) return null
    val epoch = dayIndex / daysPerEpoch
    val perm = IntArray(pool.size) { it }
    val rng = Mulberry32(simpleHash("deck-$lengthKey-$epoch"))
    for (i in pool.size - 1 downTo 1) {
        val j = (rng.nextU32() % (i + 1)).toInt()
        val tmp = perm[i]; perm[i] = perm[j]; perm[j] = tmp
    }
    val base = (dayIndex % daysPerEpoch) * dailyTotal + deal.offset
    return List(count) { pool[perm[base + it]] }
}
// ────────────────────────────────────────────────────────────────────────────

/**
 * Deterministic per-seed solution picker (mirrors Swift `generateSolutionsFromSeed`).
 * For each index, hash `"{seed}-{i}"` → `% solutionCount`; on a collision with an
 * already-used index, re-hash `"{seed}-{i}-{attempt}"` (max `solutionCount` tries).
 * `simpleHash` returns a non-negative Int, so `% solutionCount` is non-negative.
 * Requires [GameDictionary] to be initialized ([DictionaryLoader.ensureLoaded]).
 */
fun generateSolutionsFromSeed(seed: String, count: Int): List<String> {
    // Answer pool chosen by the DATE embedded in the seed (never wall clock),
    // so every client resolves the same seed identically: pre-cutover daily
    // dates → legacy list, post-cutover dailies + all non-daily seeds → curated.
    val pool = GameDictionary.solutionPool(getDailySeedDate(seed))
    // Post-DECK_CUTOVER_DATE dailies deal from the shuffled deck (§215).
    deckSolutions(seed, count, pool, DECK_DEALS_5, DECK_TOTAL_5, 5)?.let { return it }
    val solutionCount = pool.size
    val solutions = ArrayList<String>(count)
    val used = HashSet<Int>()
    for (i in 0 until count) {
        val seedWithIndex = "$seed-$i"
        var hash = simpleHash(seedWithIndex)
        var attempts = 0
        while ((hash % solutionCount) in used && attempts < solutionCount) {
            hash = simpleHash("$seedWithIndex-$attempts")
            attempts++
        }
        val index = hash % solutionCount
        used.add(index)
        solutions.add(pool[index])
    }
    return solutions
}

/** Length-specific variant (Six/Seven) — uses that length's solution pool. */
fun generateSolutionsFromSeedForLength(seed: String, count: Int, wordLength: Int): List<String> {
    // Same date-gate as the 5-letter path — pre-cutover Six/Seven dailies keep
    // their legacy words; new dailies + non-daily seeds use the curated list.
    val pool = GameDictionary.solutionPoolForLength(wordLength, getDailySeedDate(seed))
    // Six/Seven post-cutover dailies deal from their own decks (§215).
    val lengthDeals = when (wordLength) { 6 -> DECK_DEALS_6; 7 -> DECK_DEALS_7; else -> null }
    if (lengthDeals != null) {
        deckSolutions(seed, count, pool, lengthDeals, 1, wordLength)?.let { return it }
    }
    val solutionCount = pool.size
    val solutions = ArrayList<String>(count)
    val used = HashSet<Int>()
    for (i in 0 until count) {
        val seedWithIndex = "$seed-$i"
        var hash = simpleHash(seedWithIndex)
        var attempts = 0
        while ((hash % solutionCount) in used && attempts < solutionCount) {
            hash = simpleHash("$seedWithIndex-$attempts")
            attempts++
        }
        val index = hash % solutionCount
        used.add(index)
        solutions.add(pool[index])
    }
    return solutions
}

/** Random ad-hoc match seed (NOT daily): `"{epochMs}-{15-char-random}"`. */
fun generateMatchSeed(): String =
    "${System.currentTimeMillis()}-${java.util.UUID.randomUUID().toString().take(15).lowercase()}"
